go/vt/vtgate/planbuilder/builder.go

```diff
@@ -216,3 +219,111 @@ func buildVSchemaDDLPlan(stmt *sqlparser.AlterVschema, vschema ContextVSchema) (
 		AlterVschemaDDL: stmt,
 	}, nil
 }
+
+func buildFlushPlan(stmt *sqlparser.Flush, vschema ContextVSchema) (engine.Primitive, error) {
+	if len(stmt.TableNames) == 0 {
+		return buildFlushOptions(stmt, vschema)
+	}
+	return buildFlushTables(stmt, vschema)
+}
+
+func buildFlushOptions(stmt *sqlparser.Flush, vschema ContextVSchema) (engine.Primitive, error) {
+	destination, keyspace, _, err := vschema.TargetDestination("")
+	if err != nil {
+		return nil, err
+	}
+	return &engine.Send{
+		Keyspace:          keyspace,
+		TargetDestination: destination,
+		Query:             sqlparser.String(stmt),
+		IsDML:             false,
+		SingleShardOnly:   false,
+	}, nil
+}
+
+func buildFlushTables(stmt *sqlparser.Flush, vschema ContextVSchema) (engine.Primitive, error) {
+	type sendDest struct {
+		ks   *vindexes.Keyspace
+		dest key.Destination
+	}
+
+	flushStatements := map[sendDest]*sqlparser.Flush{}
+	for i, tab := range stmt.TableNames {
+		var destinationTab key.Destination
+		var keyspaceTab *vindexes.Keyspace
+		var table *vindexes.Table
+		var err error
+		table, _, _, _, destinationTab, err = vschema.FindTableOrVindex(tab)
+
+		if err != nil {
+			_, isNotFound := err.(vindexes.NotFoundError)
+			if !isNotFound {
+				return nil, err
+			}
+		}
+		if table == nil {
+			destinationTab, keyspaceTab, _, err = vschema.TargetDestination(tab.Qualifier.String())
+			if err != nil {
+				return nil, err
+			}
+			stmt.TableNames[i] = sqlparser.TableName{
+				Name: tab.Name,
+			}
+		} else {
+			keyspaceTab = table.Keyspace
+			stmt.TableNames[i] = sqlparser.TableName{
+				Name: table.Name,
+			}
+		}
+		if destinationTab == nil {
+			destinationTab = key.DestinationAllShards{}
+		}
+
+		flush, isAvail := flushStatements[sendDest{keyspaceTab, destinationTab}]
+		if isAvail {
+			flush.TableNames = append(flush.TableNames, stmt.TableNames[i])
+		} else {
+			flush = &sqlparser.Flush{
+				IsLocal:      stmt.IsLocal,
+				FlushOptions: nil,
+				TableNames:   sqlparser.TableNames{stmt.TableNames[i]},
```

Comment:
should we be using destination table name instead of original table name.
if  you have not already added a test for routed table, can you add one?

Reply:
The name is rewritten on line 270 but I'll add a case for routed tables too.